src/combineReducers.js

```diff
@@ -2,6 +2,8 @@ import { ActionTypes } from './createStore'
 import isPlainObject from 'lodash/isPlainObject'
 import warning from './utils/warning'
 
+var unexpectedKeyCache = {}
```

Comment:
I guess we could just create the cache within `combineReducers` and pass that to `getUnexpectedStateShapeWarningMessage`. Does that work for you @gaearon?